src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            IEnumerable<TSource> big = null;
+            IEnumerable<TSource> small = null;
 
-            return IntersectIterator(first, second, null);
+            if (first.Count() > second.Count())
+            {
+                big = first.Distinct();
+                small = second.Distinct();
+            }
+            else
+            {
+                big = second.Distinct();
+                small = first.Distinct();
+            }
+
+            foreach (TSource element in big)
+            {
+                if (small.Contains(element))
```

Comment:
@JonHanna You have to bear in mind that these specific Linq methods are very quick and never (not as per my knowledge) give problems. The original code adds and then removes the elements as required and through the unclear-to-me-how-optimal Set class. Just do the tests. And don't quite get what you mean with "hitting a database or web server that often".